api/proto/teleport/legacy/client/proto/authservice.proto

```diff
@@ -2242,6 +2242,31 @@ message UpdateHeadlessAuthenticationStateRequest {
   MFAAuthenticateResponse mfa_response = 3;
 }
 
+// ExportMaintenanceWindowsRequest encodes parameters for loading the
+// upgrader-facing representations of upcoming agent maintenance windows.
+message ExportMaintenanceWindowsRequest {
+  // TeleportVersion is the version of the teleport client making the request.
+  string TeleportVersion = 1;
+  // UpgraderKind represents the kind of upgrader the schedule is intended for.
+  string UpgraderKind = 2;
+}
+
+// ExportMaintenanceWindowsResponse encodes an upgrader-facing representation
+// of upcoming agent maintenance windows. Teleport agents periodically export these
+// schedules to external upgraders as part of the externally-managed upgrade system.
+message ExportMaintenanceWindowsResponse {
+  // CanonicalSchedule is the teleport-facing schedule repr.
+  types.AgentUpgradeSchedule CanonicalSchedule = 1;
+  // KubeControllerSchedule encodes upcoming upgrade upgrade windows in a format known
+  // to the kube upgrade controller. Teleport agents should treat this value as an
+  // opaque blob.
+  string KubeControllerSchedule = 2;
+  // SystemdUnitSchedule encodes the upcoming upgrade windows in a format known to
+  // the teleport-upgrade systemd unit. Teleport agents should treat this value as an
+  // opaque blob.
+  string SystemdUnitSchedule = 3;
```

Comment:
Why we bother about a client formatting ?  If a client is able to call this gRPC method is should be also to convert the `AgentUpgradeSchedule` and `SystemdUnitSchedule` proto object to the proper format on client side.

Reply:
One of the core design goals of this system was to minimize, to the greatest extent possible, the potential for agent-side bugs to prevent upgrades.  If the schedule format is bugged, we want to be able to fix it by updating the control plane _only_.

What we want to avoid is a scenario where some bug forces a user to manually upgrade because of a problem in the teleport agent code.  So long as the schedule is an opaque blob, we can be reasonably certain that the agent will either successfully export the schedule (in which case any bugs can be fixed by just updating auth), or will fail to update the schedule (which the upgraders interpret as a sign that the agent is unhealthy and should be upgraded).

If, on the other hand, the agent is the one encoding the schedule and encodes it in a bugged way (e.g. causing the updater to crash when it reads the schedule), then we might get "stuck", and need manual intervention on a per-agent basis.